src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs

```diff
@@ -875,6 +877,58 @@ internal ImmutableArray<LocalInfo<TypeSymbol>> DecodeLocalSignatureOrThrow(ref B
             }
         }
 
+        internal TypeSymbol DecodeGenericParameterConstraint(EntityHandle token, out bool isUnmanagedConstraint)
+        {
+            isUnmanagedConstraint = false;
+
+            switch (token.Kind)
+            {
+                case HandleKind.TypeSpecification:
+                    {
+                        try
+                        {
+                            var memoryReader = this.Module.GetTypeSpecificationSignatureReaderOrThrow((TypeSpecificationHandle)token);
+                            var modifiers = DecodeModifiersOrThrow(ref memoryReader, AllowedRequiredModifierType.System_Runtime_InteropServices_UnmanagedType, out var typeCode, out var modReqFound);
+                            var type = DecodeTypeOrThrow(ref memoryReader, typeCode, out _);
+
+                            if (modReqFound)
+                            {
+                                if (type.SpecialType == SpecialType.System_ValueType)
```

Comment:
>if (type.SpecialType == SpecialType.System_ValueType) [](start = 32, length = 53)

Where do we check that this is the only modifier on the type? I think we should also disallow any optional modifiers.

Reply:
Can you please explain why we should reject optional modifiers? My understanding that for any other symbol, we ignore optional modifiers we don't expect/understand. If we reject optional modifiers here, how it is different form required modifiers?

---
In reply to: [172356302](https://github.com/dotnet/roslyn/pull/24817#discussion_r172356302) [](ancestors = 172356302)

Reply:
> Can you please explain why we should reject optional modifiers? My understanding that for any other symbol, we ignore optional modifiers we don't expect/understand. If we reject optional modifiers here, how it is different form required modifiers?

First of all, we almost never completely ignore optional modifiers. We often don't complain about them, but almost never drop them on the floor. Second, read the comment that you put on line 907 below. Why is this code path different?

---
In reply to: [175279822](https://github.com/dotnet/roslyn/pull/24817#discussion_r175279822) [](ancestors = 175279822,172356302)